src/app/BufferedReadCallback.cpp

```diff
@@ -0,0 +1,256 @@
+    // write out the list - this would have removed the need for this first pass. However,
+    // we cannot actually back a TLVReader with a chained buffer since that violates the ability
+    // for us to create readers off-of readers. Each reader would assume exclusive ownership of the chained
+    // buffer and mutate the state within TLVPacketBufferBackingStore, preventing shared use.
+    //
+    // To avoid that, a single contiguous buffer is the best likely approach for now.
+    //
+    uint32_t totalBufSize = 0;
+    for (size_t i = 0; i < mBufferedList.size(); i++)
+    {
+        auto bufHandle = mBufferedList[i].Retain();
+        totalBufSize += bufHandle->TotalLength();
+    }
+
+    //
+    // Size of the start container and end container are just 1 byte each, but, let's just be safe.
+    //
+    totalBufSize += 4;
+
+    backingBuffer.Calloc(totalBufSize);
+    VerifyOrReturnError(backingBuffer.Get() != nullptr, CHIP_ERROR_NO_MEMORY);
+
+    TLV::ScopedBufferTLVWriter writer(std::move(backingBuffer), totalBufSize);
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Array, outerType));
+
+    for (auto & bufHandle : mBufferedList)
+    {
+        System::PacketBufferTLVReader reader;
+
+        reader.Init(std::move(bufHandle));
+
+        ReturnErrorOnFailure(reader.Next());
+        ReturnErrorOnFailure(writer.CopyElement(TLV::AnonymousTag, reader));
+    }
+
+    ReturnErrorOnFailure(writer.EndContainer(outerType));
+
+    writer.Finalize(backingBuffer);
+
+    aReader.Init(std::move(backingBuffer), totalBufSize);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BufferedReadCallback::AllocAndCopyElement(TLV::TLVReader & reader, System::PacketBufferHandle & handle)
+{
+    System::PacketBufferTLVWriter writer;
+    handle = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
+
+    writer.Init(std::move(handle), false);
+
+    ReturnErrorOnFailure(writer.CopyElement(TLV::AnonymousTag, reader));
+    ReturnErrorOnFailure(writer.Finalize(&handle));
+
+    // Compact the buffer down to a more reasonably sized packet buffer
+    // if we can.
+    //
+    handle.RightSize();
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BufferedReadCallback::BufferData(const ConcreteDataAttributePath & aPath, TLV::TLVReader * apData)
+{
+
+    if (aPath.mListOp == ConcreteDataAttributePath::ListOperation::ReplaceAll)
+    {
+        TLV::TLVType outerContainer;
+
+        VerifyOrReturnError(apData->GetType() == TLV::kTLVType_Array, CHIP_ERROR_INVALID_TLV_ELEMENT);
+        mBufferedList.clear();
+
+        ReturnErrorOnFailure(apData->EnterContainer(outerContainer));
+
+        CHIP_ERROR err;
+
+        while ((err = apData->Next()) == CHIP_NO_ERROR)
+        {
+            System::PacketBufferHandle handle;
+            ReturnErrorOnFailure(AllocAndCopyElement(*apData, handle));
+            mBufferedList.push_back(std::move(handle));
```

Comment:
So here is a question:  Why not push the handle and the push_back call into the helper too, and call the helper "PushCurrentElementIntoBuffer" or something?